### ISSUE TITLE
Python-Rust Otel context propagation broken

### ISSUE THREAD

**Issue by erichulburd** [rigetti/qcs-sdk-rust#545]
When we moved `py_function_sync_async` to `rigetti-pyo3`, the `opentelemetry::trace::FutureExt::with_current_context` call on the top-level future was dropped. This means the outer span with the Python OTel context was not [enter](https://tracing.rs/tracing/struct.span#method.enter)ed as the future began execution.

The temporary solution for now is to re-implement `py_function_sync_async` here and call `opentelemetry::trace::FutureExt::with_current_context` on the top-level future.

Additionally, this breakage was not caught in CI because `test-session` and `test-execution` do not run in CI. We need _some_ unit tests that can generally assert the piping here works (even if not on translation and execution themselves). 

---

In order to use the native macro from `rigetti-pyo3`, we would need that macro to match on sync functions that return a future: https://github.com/rigetti/rigetti-pyo3/issues/59 - something like the following:

```rust
py_function_sync_async!
  fn example() -> Future<()> {
    async {}.with_current_context()
  }
}
```
